### ISSUE TITLE
recomposing and merging LoCON with ONNX weights

### ISSUE THREAD

**Issue by ssube** [KohakuBlueleaf/LyCORIS#48]
I'm trying to load and blend/merge LoCON weights with the existing weights in ONNX models, and the correct math is eluding me, even after reading the Algo page a few times. The ONNX model stores weights in their full form, for example `(320, 320, 3, 3)` for a 320x320 conv2d with 3x3 kernel.

I have working code for LoRA models and the 1x1 kernel, but the `if use_cp and k_size != (1, 1)` case is giving me trouble: https://github.com/KohakuBlueleaf/LyCORIS/blob/main/lycoris/locon.py#L35

Finding the three weights, down/mid/up, is no problem but I am having trouble multiplying them back into the `(320, 320, 3, 3)` shape. Everything I've tried results in `(320, 320, 1, 1)`. For example, a model may have:

```
>>> (lora_down.shape, lora_mid.shape, lora_up.shape)
(torch.Size([4, 320, 1, 1]), torch.Size([4, 4, 3, 3]), torch.Size([320, 4, 1, 1]))
```

This is a conv2d node, with a 3x3 kernel, stride of `(2, 2)` and padding of `(1, 1, 1, 1)`.

Looking at https://github.com/KohakuBlueleaf/LyCORIS/blob/main/Algo.md#cp-decomposition, it appears that `lora_down` matches `x1` as expected, and `lora_up` would match `x2` with a `.permute((1, 0, 2, 3))`. It's the `lora_mid` or $\tau$ that is giving me trouble. This feels like it should be obvious, but how would you go about recomposing those weights?

**Comment by ssube** [comment]
After some more testing, it looks like this:

```py
weights = torch.zeros(
    (up_weight.shape[0], down_weight.shape[1], *kernel)
)
for w in range(kernel[0]):
    for h in range(kernel[1]):
        weights[:, :, w, h] = (
            up_weight.squeeze(3).squeeze(2)
            @ mid_weight[:, :, w, h]
        ) @ down_weight.squeeze(3).squeeze(2)

onnx_weights += weights * (alpha / dim)
```

produces an identical result to the merge tool, to within 1e-08.

**Comment by KohakuBlueleaf** [comment]
So there is no issue?

**Comment by ssube** [comment]
I think I have resolved the issue, if that is in fact the correct way to recompose them. It looks like you've added a lot to the merge utils, so that might answer the rest of my questions. :smile:
